### DISCUSSION TITLE
[Question] Numeric integer and decimal constraints

### DISCUSSION THREAD

**Discussion by Ultrahead** [dotnet/csharplang#2038] [Language Ideas] [answered]
There are open issues requesting a way to restrict generic types that are numbers, but is it also considered to indicate whether that numeric parameter is integer? For example to achieve something like the following pseudo-code:

```csharp
public static TNumeric GetWrappedValue<TNumeric, TInteger>(this TNumeric arg0, TInteger arg1)
   where TNumeric: struct, "numeric"
   where TInteger: struct, "integer"
{
   return arg0 % arg1;
}
```

... beaing arg1 of type byte, int, long ...

**Comment by gafter** [comment, marked as answer]
See https://github.com/dotnet/csharplang/issues/110 and https://github.com/dotnet/csharplang/issues/1711 .  I don't think we'd have language-built-in constraints for these particular things, but you could use those features to achieve that if you wanted to.

**Comment by Ultrahead** [comment]
I see. Thanks.

**Comment by amis92** [comment]
@Ultrahead does that solve the issue? Would you like to close it?

**Comment by Ultrahead** [comment]
@amis92 that depends on the final implementation of such features when available.

**Comment by Ultrahead** [comment]
Now that C# 9 is coming next, will it include a feature that allows to achieve something as shown in my initial post?

Or ...

public static TNumeric Add<TNumeric>(this TNumeric arg0, TNumeric arg1)
   where TNumeric: struct, operator+
{
   return arg0 + arg1;
}

**Comment by canton7** [comment]
No, neither shapes nor numeric constraints will be part of C# 9

**Comment by Ultrahead** [comment]
Well, let's hope they make it for C# 10, then.
